File: daily/rust/1235_maximum_profit_in_job_scheduling.rs
impl Solution {
    pub fn job_scheduling(start_time: Vec<i32>, end_time: Vec<i32>, profit: Vec<i32>) -> i32 {
        let mut jobs: Vec<(i32, i32, i32)> = start_time.iter().zip(end_time.iter()).zip(profit.iter()).map(|((s, e), p)| (*s, *e, *p)).collect();
        jobs.sort_by(|a, b| a.1.cmp(&b.1));
        let mut dp: Vec<i32> = vec![0; jobs.len()];
        dp[0] = jobs[0].2;
        for i in 1..jobs.len() {
            let mut j = i - 1;
            while j >= 0 && jobs[j].1 > jobs[i].0 {
                j -= 1;
            }
            dp[i] = std::cmp::max(dp[i - 1], dp[j] + jobs[i].2);
        }
        dp[jobs.len() - 1]
    }
}

File: daily/cpp/1235_maximum_profit_in_job_scheduling.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
 public:
  int jobScheduling(vector<int>& startTime, vector<int>& endTime,
                    vector<int>& profit) {
    int numJobs = profit.size();  // length of jobs
    vector<tuple<int, int, int>> jobs(
        numJobs);  // tuple of (start, end, profit)

    for (int i = 0; i < numJobs; ++i) {
      jobs[i] = make_tuple(endTime[i], startTime[i], profit[i]);
    }

    sort(jobs.begin(), jobs.end());
    vector<int> dp(numJobs +
                   1);  // dp[i] = max profit from jobs[0] to jobs[i - 1]

    for (int i = 0; i < numJobs; ++i) {
      auto [endTime, startTime, profit] = jobs[i];

      int latestNonConflictJobIndex =
          upper_bound(jobs.begin(), jobs.begin() + i, startTime,
                      [&](int time, const auto& job) -> bool {
                        return time < get<0>(job);
                      }) -
          jobs.begin();

      dp[i + 1] = max(dp[i], dp[latestNonConflictJobIndex] + profit);
    }

    return dp[numJobs];
  }
};